src/VersionUpdater/SelfUpdate.php

```diff
@@ -0,0 +1,192 @@
+<?php
+
+/*
+ * This file is part of Psy Shell.
+ *
+ * (c) 2012-2022 Justin Hileman
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Psy\VersionUpdater;
+
+use Psy\Exception\ErrorException;
+use Psy\Shell;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Self update command.
+ *
+ * If a new version is available, this command will download it and replace the currently installed version
+ */
+class SelfUpdate
+{
+    const URL_PREFIX = 'https://github.com/bobthecow/psysh/releases/download';
+    const SUCCESS = 0;
+    const FAILURE = 1;
+
+    /** @var Checker */
+    private $checker;
+
+    /** @var Installer */
+    private $installer;
+
+    /** @var Downloader */
+    private $downloader;
+
+    public function __construct(Checker $checker, Installer $installer)
+    {
+        $this->checker = $checker;
+        $this->installer = $installer;
+    }
+
+    /**
+     * Allow the downloader to be injected for testing.
+     *
+     * @param Downloader $downloader
+     *
+     * @return void
+     */
+    public function setDownloader(Downloader $downloader)
+    {
+        $this->downloader = $downloader;
+    }
+
+    /**
+     * Get the currently set Downloader or create one based on the capabilities of the php environment.
+     *
+     * @return Downloader
+     *
+     * @throws ErrorException if a downloader cannot be created for the php environment
+     */
+    private function getDownloader(): Downloader
+    {
+        if (!isset($this->downloader)) {
+            return Downloader\Factory::getDownloader();
+        }
+
+        return $this->downloader;
+    }
+
+    /**
+     * Build the download URL for the latest release.
+     *
+     * The file name used in the URL will include the flavour postfix extracted from the current version
+     * if it's present
+     *
+     * @param string $latestVersion
+     *
+     * @return string
+     */
+    private function getAssetUrl(string $latestVersion): string
+    {
+        $versionPostfix = '';
+        if (\strpos(Shell::VERSION, '-')) {
+            $versionPostfix = \substr(Shell::VERSION, \strpos(Shell::VERSION, '-'));
```

Comment:
oh, sorry i missed this before. the version number looks like this: `v0.11.8-64-ga0c5d077+php70`. this is:

```
{version}-{git tag annotation}+{build flavor}
```

this lines up with semver's

```
{version}-{pre-release}+{build}
```

… unfortunately i don't name my release tarballs that way, so we need to compare/extract in one format and download in another 😬

Reply:
Oh, my bad - I see I have the format of `Shell::VERSION` wrong when a build-flavor is present `{version}+{build}` rather than `{version}-{build}` and the url should always use `{version}-{build}`. 

I'm not sure I understand how the `-{pre-release}` part fits in the picture though. Does that need to be handled when using the `Shell::VERSION` value here? Or is the issue simply dealing with the `+` vs `-` in the case the version includes the build flavor?

Reply:
Oh, you can ignore the pre-release bit. I was just explaining why you might have seen a `-*` in a version string. All we need to account for here is whatever comes after a `+`.